### PR TITLE
[ENG-633] Add registration metrics gathering script

### PR DESCRIPTION
**Pull request by brianjgeiger**
## Purpose

Gather metrics for metascience to measure Registration Usage per the [requirements](https://docs.google.com/document/d/12b4PgO7nZkjgGaEk5qV-MSJBsThUomRvUOU087A7DOI/edit#heading=h.13dsofnworfu)

## Changes

- Add a management command to gather the metrics
- Add defaults for being able to run the command with celery beat and also to upload data to OSF

## Devops

This will require settings for an OSF token (can be the same as the data storage usage token) and an OSF folder to hold the info 
```
REG_METRICS_BASE_FOLDER = 'https://files.osf.io/v1/resources/r83uz/providers/osfstorage/5d0bdac73466a6001a93fc17/'
```

## QA Notes

This is not expected to need QA

## Documentation

No documentation

## Side Effects

It's a quick script (should take less than a minute to run) and shouldn't cause problems

## Ticket

https://openscience.atlassian.net/browse/ENG-633

### CODE REVIEW

File: osf/management/commands/registration_schema_metrics.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+/*  get all top-level, non-deleted, registrations that were registered or retracted last month -
+some approved retraction don't have retraction dates..why? */
+
+WITH toplevel_regs AS (SELECT DISTINCT ON (osf_abstractnode.root_id) CAST(osf_abstractnode.registered_date AS DATE),
+CAST(osf_retraction.date_retracted AS DATE),
+        json_object_keys (registered_meta::json)
+        FROM osf_abstractnode
+        LEFT JOIN osf_retraction
+        ON osf_abstractnode.retraction_id = osf_retraction.id
+        WHERE osf_abstractnode.type LIKE '%osf.registration%'
```
[the remark this answers is not in the document]
Comment (brianjgeiger) [reply osf/management/commands/registration_schema_metrics.py]: Yup

File: osf/management/commands/registration_schema_metrics.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        AND ((osf_abstractnode.registered_date >= date_trunc('month', current_date - interval '1' month)
+            AND osf_abstractnode.registered_date < date_trunc('month', current_date))
+            OR (osf_retraction.date_retracted >= date_trunc('month', current_date - interval '1' month)
+            AND osf_retraction.date_retracted < date_trunc('month', current_date)))
+            AND osf_abstractnode.is_deleted IS FALSE
+        GROUP BY osf_abstractnode.root_id, CAST(osf_abstractnode.registered_date AS DATE),
```
[the remark this answers is not in the document]
Comment (csoderberg) [reply osf/management/commands/registration_schema_metrics.py]: Yep. I guess I could also filter out nodes where the node id doesn't equal the root.id to only keep top level registrations. Let me know if you have a preference

[the remark this answers is not in the document]
Comment (pattisdr) [reply osf/management/commands/registration_schema_metrics.py]: this is probably fine!

File: osf/management/commands/registration_schema_metrics.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        json_object_keys (registered_meta::json)
+        FROM osf_abstractnode
+        LEFT JOIN osf_retraction
+        ON osf_abstractnode.retraction_id = osf_retraction.id
+        WHERE osf_abstractnode.type LIKE '%osf.registration%'
+        AND ((osf_abstractnode.registered_date >= date_trunc('month', current_date - interval '1' month)
+            AND osf_abstractnode.registered_date < date_trunc('month', current_date))
+            OR (osf_retraction.date_retracted >= date_trunc('month', current_date - interval '1' month)
+            AND osf_retraction.date_retracted < date_trunc('month', current_date)))
+            AND osf_abstractnode.is_deleted IS FALSE
```
[the remark this answers is not in the document]
Comment (csoderberg) [reply osf/management/commands/registration_schema_metrics.py]: I'm pretty sure yes. We have deleted registrations that are recent and not withdrawn, which would suggest they were canceled. Also, I tracked down some of the logs for projects that had deleted and non-deleted registrations and the deleted registration times match up with logs about rejecting registrations.

File: osf/management/commands/registration_schema_metrics.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    filename = 'form_types_{}.csv'.format(last_month.strftime('%Y-%m'))
+
+    with connection.cursor() as cursor:
+        cursor.execute(REGISTRATION_METRICS_SQL)
+        if dry_run:
+            for row in cursor.fetchall():
+                row_to_write = []
+                for s in row:
+                    item = s.encode('utf-8') if isinstance(s, (str, unicode)) else s
+                    row_to_write.append(item)
```
[the remark this answers is not in the document]
Comment (brianjgeiger) [reply osf/management/commands/registration_schema_metrics.py]: Yeah, that would make sense.

File: website/settings/defaults.py [hunk 5]
```diff
@@ -566,6 +566,11 @@ class CeleryConfig:
                 'schedule': crontab(minute=0, hour=7),  # Daily 2:00 a.m.
                 'kwargs': {'dry_run': False}
             },
+            'registration_schema_metrics': {
+                'task': 'management.commands.registration_schema_metrics',
+                'schedule': crontab(minute=45, hour=7, day_of_month=3),  # Third day of month 2:45 a.m.
+                'kwargs': {'dry_run': False}
+            },
```
Comment (pattisdr) [inline website/settings/defaults.py]: Will defer to @mfraezz to make sure the timing of this is suitable!

**Comment by pattisdr** [comment]
merging to develop, no QA required.

**Pull request merged**